### PR TITLE
Singletone homework_1

### PR DESCRIPTION
**Pull request by 3dgarr**

### CODE REVIEW

File: Singleton/version_2/Singleton.hpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# include <iostream>
+# include <mutex>
+# include <thread>
+
+class Singleton
+{
+	public:
+		void				setName(std::string const&);
+		std::string			getName() const;
+		static Singleton*	getInstacne(std::string const&);
```
Comment (EdoHarutyunyan) [inline Singleton/version_2/Singleton.hpp:13]:
```suggestion
		static Singleton*	getInstance(std::string const&);
```

File: Singleton/version_2/Singleton.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+Singleton::~Singleton()
+{
+	std::cout << __PRETTY_FUNCTION__ << std::endl;
+}
+
+Singleton*	Singleton::getInstacne(std::string const& n)
+{
+	_mutex.lock();
+	if (!_instance)
```
Comment (EdoHarutyunyan) [inline Singleton/version_2/Singleton.cpp:34]: It is recommended to explicitly check for `nullptr`

File: Singleton/version_2/Singleton.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+std::unique_ptr<Singleton>	Singleton::_instance = nullptr;
+std::mutex					Singleton::_mutex;
+
+Singleton::Singleton(std::string const & n)
+	:	_name(n)
+{
+	std::cout << __PRETTY_FUNCTION__ << std::endl;
+}
+
+void	Singleton::setName(std::string const& n)
```
Comment (EdoHarutyunyan) [inline Singleton/version_2/Singleton.cpp:12]: Please choose more descriptive names (n->name)

File: Singleton/version_1/Singleton.hpp [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+	public:
+		Singleton(Singleton const&) = delete;
+		Singleton(Singleton&&) = delete;
+		Singleton& operator=(Singleton const&) = delete;
+		Singleton& operator=(Singleton&&) = delete;
+		~Singleton();
+
+	private:
+		std::string _name;
```
Comment (EdoHarutyunyan) [inline Singleton/version_1/Singleton.hpp:27]: Just for your info:

In C++ it is accepted to have members with `m_` prefix.
`_` prefix is more usable in languages like C#.

File: Singleton/version_1/Singleton.hpp [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+		// void				setName(std::string const&);
+		std::string			getName() const;
+		static Singleton&	getInstacne(std::string const&);
+	
+	private:
+		Singleton(std::string const&);
+
+	public:
+		Singleton(Singleton const&) = delete;
+		Singleton(Singleton&&) = delete;
```
Comment (EdoHarutyunyan) [inline Singleton/version_1/Singleton.hpp:21]: Please check this article: [link](https://stackoverflow.com/questions/23771194/why-should-i-delete-move-constructor-and-move-assignment-operator-in-a-singleton#:~:text=If%20you%20declare%20a%20copy%20constructor%20(even%20if%20you%20define%20it%20as%20deleted%20in%20the%20declaration)%2C%20no%20move%20constructor%20will%20be%20declared%20implicitly.%20Cf.%20C%2B%2B11%2012.8/9%3A)

The interesting part is that when you delete a copy ctor you are actually "declare" it :)

File: Singleton/version_2/main.cpp [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	const std::string	name("David");
+
+    Singleton* singleton = Singleton::getInstacne(name);
+	std::cout << singleton->getName() << " -- thread1\n";
+
+}
+
+void thread2()
+{
+	const std::string	name("Eduard");
```
Comment (EdoHarutyunyan) [inline Singleton/version_2/main.cpp:14]: It is better to use curly braces whenever it is possible. You can take it as a rule:
```
const std::string name{"Eduard"};
```

File: Singleton/version_1/Singleton.cpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+std::string Singleton::getName() const
+{
+	return _name;
+}
+
+Singleton&	Singleton::getInstacne(std::string const& n)
+{
+	static Singleton instance(n);
+
+	return	(instance);
```
Comment (EdoHarutyunyan) [inline Singleton/version_1/Singleton.cpp:25]:
```suggestion
	return instance;
```

File: Singleton/version_1/Singleton.hpp (hunk 5 above)
Comment (3dgarr) [reply Singleton/version_1/Singleton.hpp:21]: Thank you for information, that was useful!
I'll follow your advice :)

File: Singleton/version_1/Singleton.hpp (hunk 4 above)
Comment (3dgarr) [reply Singleton/version_1/Singleton.hpp:27]: Thanks for this too. I'll follow for this style, I like it

File: Singleton/version_2/Singleton.cpp (hunk 3 above)
Comment (3dgarr) [reply Singleton/version_2/Singleton.cpp:12]: Accepted 🤐

File: Singleton/version_2/Singleton.cpp (hunk 2 above)
Comment (3dgarr) [reply Singleton/version_2/Singleton.cpp:34]: Should I write like this?
_instance == nullptr !?

File: Singleton/version_2/main.cpp (hunk 6 above)
Comment (3dgarr) [reply Singleton/version_2/main.cpp:14]: Accepted 🤐

File: Singleton/version_2/Singleton.cpp (hunk 2 above)
Comment (EdoHarutyunyan) [reply Singleton/version_2/Singleton.cpp:34]: yes

**Pull request merged**